src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -184,7 +184,13 @@ STATIC_ASSERT(max_yellow_zone <= max_red_zone);
 #define LOG_ZONES(...) log_debug( CTRL_TAGS )(__VA_ARGS__)
 
 static size_t buffers_to_cards(size_t value) {
-  return value * G1UpdateBufferSize;
+  if (value == 0) return 0;
+  size_t res = value * G1UpdateBufferSize;
+
+  if (res < value || res < G1UpdateBufferSize) { // Check overflow
+    vm_exit_during_initialization("buffers_to_cards overflow!");
+  }
+  return res;
```

Comment:
buffers_to_cards is only used with command line options as arguments. So
it's not inappropriate to use vm_exit_during_initialization, even though
this function isn't obviously about initialization. Maybe it should be
changed to

// Convert configuration values in units of buffers to number of cards.
static size_t configuration_buffers_to_cards(size_t value, const char* value_name) {
  ... use value_name in the error message
}

An alternative would be to move the checking to constraint functions for
the command line options. That puts the checking far from the code where
it matters though, but maybe a comment in the conversion function referring
to the option constraints would be sufficient. This might provide better
reporting?